CalibrationTablesBuilder

jean-philippe-martin opened the pull request:
This code wraps around BaseRecalibrator and presents a very basic interface (set up, add reads, teardown) that's going to be used at each Dataflow worker.

Challenges here:
(1) I need to convert the intervals to Features because that's what the BaseRecalibrator class uses, and SimpleInterval is not a subclass of Feature. This may change in the future.
(2) BaseRecalibrator takes Features as inputs - the only simple Feature class I found I could reuse is ArtificialTestFeature. Please let me know if there is a better choice (solving (1) also solves this)
(3) I didn't find code to test overlap between a SimpleInterval and a Feature. Rather than roll my own I chose to use the SimpleInterval overlap test and convert to Feature lazily instead of eagerly. This may cause an interval to be converted more than once.

So please consider this the start of a discussion on "here is something that works, but surely there's a better way?" I'm not so much looking for every performance opportunity, but ideally I'd like to avoid using ArtificialTestFeature if a better candidate is available.

droazen commented:
Hi @jean-philippe-martin ,

A `Feature` in our codebase has a specific meaning that is different from "interval": it is a record that 1. has a location on the genome plus (typically) some metadata information about that location and 2. is in one of the formats supported by our file-parsing framework tribble and is the product of a tribble codec. A VCF record is an example of a `Feature`. 

The common interface between `Feature` and `SimpleInterval` is called `Locatable`.

I recommend (for now) that you simply alter your uprooted version of BQSR to take a `List<Locatable>` instead of a `List<Feature>` in `apply()`. This should require no code changes beyond changing method parameter types, and it will allow you to feed BQSR `SimpleIntervals` for the known sites for now, and `Features` like VCF records later on when we're ready for that.

Please do return `ArtificialTestFeature` to the `FeatureDataSourceUnitTest` from which it came -- this is a very incomplete class meant only for testing purposes and not for external use.

jean-philippe-martin commented:
@droazen, thank you! I've switched to Locatable and reverted ArtificialTestFeature. Now we need one fewer translation!

droazen commented on src/main/java/org/broadinstitute/hellbender/dev/pipelines/bqsr/CalibrationTablesBuilder.java:
```diff
@@ -0,0 +1,93 @@
+ * (OK, yes, I need the reference genome too).
+ *
+ * This class wraps around BaseRecalibratorUprooted, giving it a cozy environment outside of the normal
+ * CommandLine framework. It's also completely independent of Dataflow, so it could be tested outside of that framework.
```
Since you admit it could be tested independently, why not add unit tests for it now as part of this pull request? :)

droazen commented on src/main/java/org/broadinstitute/hellbender/dev/pipelines/bqsr/CalibrationTablesBuilder.java:
```diff
@@ -0,0 +1,93 @@
+
+  private BaseRecalibratorUprooted br;
+  private SAMFileHeader header;
+  private ReferenceDataSource reference;
```
Make these members final.

droazen replied:
(except `br`, which I see you set to null on traversal end)

droazen commented on src/main/java/org/broadinstitute/hellbender/dev/pipelines/bqsr/CalibrationTablesBuilder.java:
```diff
@@ -0,0 +1,93 @@
+    ret[1] = fastaFilename + ".fai";
+    int lastDot = fastaFilename.lastIndexOf('.');
+    ret[2] = fastaFilename.substring(0, lastDot) + ".dict";
+    return ret;
```
This `String[]` return value is ugly -- could you refactor this into two static utility methods `getFastaIndexFileName(fastaFileName)` and `getFastaDictionaryFileName(fastaFileName)`, and put these in a new utility class `utils/reference/ReferenceUtils`?

droazen commented on src/main/java/org/broadinstitute/hellbender/dev/pipelines/bqsr/CalibrationTablesBuilder.java:
```diff
@@ -0,0 +1,93 @@
+     * The skipIntervals have to be 1-based, close-ended.
+     */
+  public void add(Iterable<Read> reads, List<SimpleInterval> skipIntervals) {
+    if (null==br) throw new RuntimeException("Can't call add after done");
```
Make this a `GATKException`

akiezun commented:
This code comes with no tests and it duplicates the code we have in BQSR, which is not the right way to do it I think. 

I think there are 2 ways to proceed. One is to refactor the current BQSR code to use an abstraction like the Recalibration Table builder (which may not be a bad idea by itself) - then you can reuse existing tests. Another way it to treat this a from-scratch reimplementation - but then we need all this new code fully tested.

back to @jean-philippe-martin

droazen commented:
Review complete

droazen commented:
@akiezun These changes are not compatible with the walker interface / engine, and should not be merged into the walker version of BQSR, in my opinion. Eg., the only non-trivial code here finds the skip intervals that overlap the current read, which is something the walker engine already does for you (via the `FeatureContext` passed to `apply()`).

droazen commented on src/main/java/org/broadinstitute/hellbender/dev/pipelines/bqsr/CalibrationTablesBuilder.java:
```diff
@@ -0,0 +1,93 @@
+    for (Read r : reads) {
+      SAMRecord sr = GenomicsConverter.makeSAMRecord(r, header);
+      final SimpleInterval readInterval = sr.getReadUnmappedFlag() ? null :
+          new SimpleInterval(sr.getReferenceName(), sr.getAlignmentStart(), sr.getAlignmentEnd());
```
Use the `SimpleInterval` constructor that takes a `Locatable` (since `SAMRecord` implements `Locatable`)

droazen commented:
JP & I had a discussion over email about a path by which he could add tests to this branch and show that he matches the recalibration table output by the walker BQSR for a given set of input(s).

jean-philippe-martin commented:
Addressed feedback, updated so it can write reports and added test that checks the exact same reports as the BaseRecalibrator integration test (they pass).

akiezun commented:
LGTM. small stylistic comments. squash, rebase and feel free to merge when tests pass.
after rebasing you will find that StandardCovariateList is your friend

back to @jean-philippe-martin

jean-philippe-martin commented:
Thank you for the feedback! Will update style.

jean-philippe-martin commented:
Done. Squashed, rebased, tested, merged.
